### ISSUE TITLE
How can i swap faces between two images?

### ISSUE THREAD

**Issue by alan-ai-learner** [zyainfal/One-Shot-Face-Swapping-on-Megapixels#26]
I have two images and i want to swap the face from one image to another. Any suggestion would be great?
thanks

**Comment by zyainfal** [comment]
You need to get face segmentations at the first step, please refer to [this](https://github.com/zllrunning/face-parsing.PyTorch) for details.
Then you can swap the two faces by giving the path to the inference code, or you can read images by `cv2.imread()` directly.

**Comment by alan-ai-learner** [comment]
Hi, thankyou so much for replying, i tried the way you said, but there is a
bug in the inference.py file, op module is not found , I tried to downgrade
the pytorch but the version 1.3 is not found that you used.any help would
be appreciated.
Thanks

**Comment by zyainfal** [comment]
`op module` is from StyleGAN2.
Please refer to [stylegan2-pytorch](https://github.com/rosinality/stylegan2-pytorch) for details. (please take care of CUDA version, it is crucial to compile op modules)

**Comment by alan-ai-learner** [comment]
Thanks, I've managed to solved it, but I'm facing this new error, what is exactly and source and target id, I'm trying to replace whole face here, not an particular face attribute.
![err](https://user-images.githubusercontent.com/53686128/142863092-d59d406e-2de7-4bcc-8f3d-a88a7b9918ee.png)

**Comment by zyainfal** [comment]
I assume you give the image name to args.
However, this `inference.py` is coded for CelebA-HQ, where srcID and tgtID are the image indexes.
So in your case, you can read images and their segmentations by `cv.imread()` directly to `src_face`, `tgt_face`, and `tgt_mask  `.

**Comment by alan-ai-learner** [comment]
ok, thanks, will this image swapping works on the images like a profile pictures, like in the Celeb HQ, the images are mainly containg the faces of the celebrities, and i want to swap faces of an image which is like this
![whole_img](https://user-images.githubusercontent.com/53686128/142865434-30d221a2-ad1d-4e73-b38f-482c650bbc4e.png)

**Comment by zyainfal** [comment]
For this kind of image, you need to detect and align the face according to the CelebA-HQ paper (or simply by RetinaFace). Then, you can swap them if everything goes well.

**Comment by alan-ai-learner** [comment]
> You need to get face segmentations at the first step, please refer to [this](https://github.com/zllrunning/face-parsing.PyTorch) for details. Then you can swap the two faces by giving the path to the inference code, or you can read images by `cv2.imread()` directly.

here ive to find the segmentation of target image or the sorce images.
![Uploading 000109.jpg…](Source Img)
![Uploading 000109.jpg…](Segmentation of source img)
![45721bd4-aeca-4191-b0fc-8d717b6d1c3f](https://user-images.githubusercontent.com/53686128/142866777-84d60af0-eb31-42de-81cd-79d5fbe2be07.jpg)

**Comment by zyainfal** [comment]
The target segmentation.

**Comment by alan-ai-learner** [comment]
> For this kind of image, you need to detect and align the face according to the CelebA-HQ paper (or simply by RetinaFace). Then, you can swap them if everything goes well.

Can you suggest some repo that were doing it, that will be great help.
Thanks

**Comment by alan-ai-learner** [comment]
> The target segmentation.
Again the same,
![ftm](https://user-images.githubusercontent.com/53686128/142867721-94922a30-0aa4-4a0b-b8ac-d2b63dfedff4.jpg)

**Comment by zyainfal** [comment]
You may make mistake of the target segmentation, it should have values from 0 to 17 (the semantic index), not RGB values.
Just like in https://github.com/zllrunning/face-parsing.PyTorch/blob/master/evaluate.py line 39,
save it as segmentaion image.

**Comment by alan-ai-learner** [comment]
> You may make mistake of the target segmentation, it should have values from 0 to 17 (the semantic index), not RGB values. Just like in https://github.com/zllrunning/face-parsing.PyTorch/blob/master/evaluate.py line 39, save it as segmentaion image.

yeah, i need to use the png segementation, getting this after, seem not realistic
![ftm (1)](https://user-images.githubusercontent.com/53686128/142868376-97e4dbb0-8a8b-4c50-a57d-e719471059b5.jpg)

**Comment by zyainfal** [comment]
Well, this is the result. You can try `injection` model to check which is better.

**Comment by alan-ai-learner** [comment]
> > For this kind of image, you need to detect and align the face according to the CelebA-HQ paper (or simply by RetinaFace). Then, you can swap them if everything goes well.
> 
> Can you suggest some repo that were doing it, that will be great help. Thanks

Okay any suggeston on this

**Comment by zyainfal** [comment]
A simple practice is to use RetinaFace to detect and align the face.
RetinaFace: https://github.com/deepinsight/insightface/tree/master/detection/retinaface
Align functioin: https://github.com/deepinsight/insightface/blob/ce3600a74209808017deaf73c036759b96a44ccb/recognition/arcface_mxnet/common/build_eval_pack.py Line 72 get_norm_crop()

**Comment by alan-ai-learner** [comment]
Thanks, I'll check them out.

**Comment by an-zi-li** [comment]
> 

Hello, may I ask how to get the mask that can be used for human face? I would be grateful if you could reply.

**Comment by an-zi-li** [comment]
> > You need to get face segmentations at the first step, please refer to [this](https://github.com/zllrunning/face-parsing.PyTorch) for details. Then you can swap the two faces by giving the path to the inference code, or you can read images by `cv2.imread()` directly.
> 
> here i've to find the segmentation of target image or the source images. getting the following result source Img: ![000109](https://user-images.githubusercontent.com/53686128/142867095-a118fc7d-042c-40ef-92cf-374b6d3d093a.jpg) Source IMg Segmentation: ![000109](https://user-images.githubusercontent.com/53686128/142867290-0b7f0597-7387-4761-876b-0f10f2e53403.jpg)
> 
> Target IMg: ![000735](https://user-images.githubusercontent.com/53686128/142867153-71447406-037a-49b2-b327-985d65a4445e.jpg) Output Img: ![45721bd4-aeca-4191-b0fc-8d717b6d1c3f](https://user-images.githubusercontent.com/53686128/142867194-15a8b695-8c2b-47f3-b122-ab6113240572.jpg)

Hello, may I ask how to get the mask that can be used for human face? I would be grateful if you could reply.

**Comment by zyainfal** [comment]
> You may make mistake of the target segmentation, it should have values from 0 to 17 (the semantic index), not RGB values. Just like in https://github.com/zllrunning/face-parsing.PyTorch/blob/master/evaluate.py line 39, save `vis_parsing_anno ` as segmentaion image.

Have you tried this?

**Comment by an-zi-li** [comment]
> > 你可能会弄错目标分割，它应该有从 0 到 17 的值（语义索引），而不是 RGB 值。就像在https://github.com/zllrunning/face-parsing.PyTorch/blob/master/evaluate.py第 39 行中一样，保存`vis_parsing_anno `为分段图像。
> 
> 你试过这个吗？

![image](https://user-images.githubusercontent.com/95530998/158289483-6ce680e0-13f5-41c0-860d-ef3bc2061ac6.png)
So far I have got the whole face mask, so how do I divide each part
